runtime/runtime-params-estimator/costs.txt

```diff
@@ -36,6 +36,8 @@ Keccak512Byte                                      12_216_567
 Ripemd160Base                                     284_558_362
 Ripemd160Block                                    226_702_528
 EcrecoverBase                                  92_940_662_819
+Ed25519VerifyBase                           3_365_369_625_000
+Ed25519VerifyByte                                 291_580_479
```

Comment:
@blasrodri fyi these numbers don't look right to me. It means that we can do at most ~300 signature verifications per chunk. According to https://github.com/dalek-cryptography/ed25519-dalek#benchmarks, we should be able to do two orders of magnitude more. cc @jakmeier

Reply:
I'd like to get input from @jakmeier here.

Reply:
@blasrodri  Well, as I asked before, did you actually run these two estimations? It seems you just copied the values from `wasm_ecrecover_base` and `wasm_utf8_decoding_byte`, or you mixed them up somehow. But it's basically impossible that you get these exact values from the estimation.

Reply:
```
    ~/c/nearcore/runtime/runtime-params-estimator    blas-impleme…st-functions  cargo run --release --package runtime-params-estimator --features required,nightly --bin runtime-params-estimator -- --accounts-num 200 --additional-accounts-num 2000 --iters 1 --warmup-iters 1 --metric time --costs Ed25519VerifyBase,Ed25519VerifyByte
    Finished release [optimized] target(s) in 0.17s
     Running `/home/blas/composable-finance/nearcore/target/release/runtime-params-estimator --accounts-num 200 --additional-accounts-num 2000 --iters 1 --warmup-iters 1 --metric time --costs Ed25519VerifyBase,Ed25519VerifyByte`
[elapsed 00:00:00 remaining 00:00:00] Writing into storage ████████████████████    2000/2000   
Ed25519VerifyBase                                   43_613_893_904 gas [                 43.614µs] UNCERTAIN (computed in 2.45s) HIGH-VARIANCE: runtime/runtime-params-estimator/src/utils.rs:300:18
Ed25519VerifyByte                                      456_577_052 gas [                    457ns] UNCERTAIN (computed in 1.97s) HIGH-VARIANCE: runtime/runtime-params-estimator/src/utils.rs:300:18
```

Does this makes sense?

Reply:
Hm, it's higher a lot higher than what we were expecting. Can you try to fully optimize the binary by setting these environment variables?
```
export CARGO_PROFILE_RELEASE_CODEGEN_UNITS=1;
export CARGO_PROFILE_RELEASE_LTO=fat;
```
Also, please run with more iterations. For example `--iters 5 --warmup-iters 1`. Hopefully the uncertain warning will go away then.

Reply:
not sure but when I change `iters > 1` I never get the result. Tried with 5 and 2. With 1 I get it right away

Reply:
That's very odd. With a total of two iterations it took 2.45s on your machine. Adding more iterations is doing the same work just a few more times. So with a total of 6 iterations, I would expect it to be no more than 7.5s.

I don't really have the capacity to debug this myself. But it sounds like maybe there is a bug somewhere in the new estimation code?

Reply:
same when I increase `warmup-iters`

Reply:
This is what I get with iters = 1 and binary optimized

```
     Running `/home/blas/composable-finance/nearcore/target/release/runtime-params-estimator --accounts-num 200 --additional-accounts-num 2000 --iters 1 --warmup-iters 1 --metric time --costs Ed25519VerifyBase,Ed25519VerifyByte`
[elapsed 00:00:00 remaining 00:00:00] Writing into storage ████████████████████    2000/2000   
Ed25519VerifyBase                                   40_311_888_867 gas [                 40.312µs] UNCERTAIN (computed in 2.28s) HIGH-VARIANCE: runtime/runtime-params-estimator/src/utils.rs:300:18
Ed25519VerifyByte                                      423_978_605 gas [                    424ns] UNCERTAIN (computed in 1.84s) HIGH-VARIANCE: runtime/runtime-params-estimator/src/utils.rs:300:18
```

Reply:
Didn't change much. But I just realised I didn't read `Ed25519VerifyBase` properly. Actually, these gas costs are not too bad.

From my point of view we can go with these numbers as far as this PR is concerned. We still have to fine-tune them by running on standardised hardware before stabilising the feature. But this should happen after merging the basic feature.

Reply:
Oh, but we still need to figure out why estimations don't terminate with more than 2 iterations.

Reply:
I'd start by trying to find out where exactly it hangs.

We know that all existing estimations in master are running fine no matter how many iterations, we even have a daily job for it. So it is almost certainly a bug in the newly added code. Some well-placed `println!` should be sufficient to find out where it starts hanging.

Reply:
It's hanging there:
```diff
diff --git a/runtime/runtime-params-estimator/src/utils.rs b/runtime/runtime-params-estimator/src/utils.rs
index b1036f64b..76224c8f3 100644
--- a/runtime/runtime-params-estimator/src/utils.rs
+++ b/runtime/runtime-params-estimator/src/utils.rs
@@ -54,8 +54,10 @@ pub(crate) fn transaction_cost_ext(
         let mut blocks = Vec::with_capacity(n_blocks);
         for _ in 0..n_blocks {
             let mut block = Vec::with_capacity(block_size);
-            for _ in 0..block_size {
+            for j in 0..block_size {
+                println!("inner loop {}", j);
                 let tx = make_transaction(testbed.transaction_builder());
+                println!("inner loop {}", j);
                 block.push(tx)
             }
             blocks.push(block)
```
Not sure how I can further debug it

Reply:
I'm getting the same issue when running
```
 cargo run --release --package runtime-params-estimator --features required,nightly --bin runtime-params-estimator -- --accounts-num 200 --additional-accounts-num 2000 --iters 3 --warmup-iters 0 --metric time --costs EcrecoverBase
```

Reply:
Oh, now I see the problem^^

A pool of only 200 accounts is too small, as the way the estimation is written right now, it tries to use only unique accounts. Sorry about that, I will file a bug for it, it'd not something that you need to fix in this PR. In the meantime, please run with something like `--accounts-num 20000 --additional-accounts-num 20000`.

For context, usually we run it with millions of accounts to simulate realistic DB latencies. But for this estimation, the DB should not matter, therefore I gave you an example with fewer accounts to make it faster. Apparently it was a too small.

Reply:
Filed #7448

Thank you @blasrodri for helping out with the debugging on this!